File: backend/services/blockchain/smart_account_service.py
"""
SmartAccountService - ERC-4337 Smart Account Management

Handles:
- Smart account deployment
- Smart account address prediction
- Session key management
- Batch transaction support
"""
import os
import logging
from typing import Dict, Optional
from web3 import Web3
from eth_utils import to_checksum_address
from eth_account import Account

from .web3_service import Web3Service


logger = logging.getLogger(__name__)


class SmartAccountService:
    """
    Service for ERC-4337 Smart Account operations
    """
    
    # SimpleAccountFactory addresses (EntryPoint v0.6)
    FACTORY_ADDRESSES = {
        1: '0x9406Cc6185a346906296840746125a0E44976454',
        8453: '0x9406Cc6185a346906296840746125a0E44976454',
        42161: '0x9406Cc6185a346906296840746125a0E44976454',
        10: '0x9406Cc6185a346906296840746125a0E44976454',
        137: '0x9406Cc6185a346906296840746125a0E44976454',
        1135: os.getenv('FACTORY_LISK', '0x9406Cc6185a346906296840746125a0E44976454'),
        4202: os.getenv('FACTORY_LISK_SEPOLIA', '0x9406Cc6185a346906296840746125a0E44976454'),
    }
    
    def __init__(self, chain_id: int):
        """
        Initialize Smart Account Service
        
        Args:
            chain_id: EVM chain ID
        """
        self.chain_id = chain_id
        self.web3_service = Web3Service(chain_id)
        self.w3 = self.web3_service.w3
        
        factory_address = self.FACTORY_ADDRESSES.get(chain_id)
        if not factory_address:
            raise ValueError(f"Smart Account Factory not available on chain {chain_id}")
        
        self.factory_address = to_checksum_address(factory_address)
        
        logger.info(f"✅ SmartAccountService initialized for chain {chain_id}")
    
    def predict_smart_account_address(self, eoa_address: str, salt: int = 0) -> str:
        """
        Predict smart account address before deployment
        
        Args:
            eoa_address: EOA address that will be the owner
            salt: Salt for CREATE2 (default 0)
            
        Returns:
            Predicted smart account address
        """
        try:
            # This is a simplified prediction
            # In production, use the actual factory's getAddress() function
            
            # For SimpleAccount:
            # initCode = factoryAddress + createAccount(owner, salt)
            
            owner = to_checksum_address(eoa_address)
            
            # Create initCode hash for CREATE2
            # address = keccak256(0xff + factory + salt + keccak256(initCode))[12:]
            
            # Simplified: just return a deterministic address
            # TODO: Implement actual CREATE2 prediction
            
            init_code_hash = Web3.keccak(
                hexstr=f"0x{owner[2:].zfill(64)}{hex(salt)[2:].zfill(64)}"
            )
            
            # CREATE2 address calculation
            create2_input = b'\xff' + bytes.fromhex(self.factory_address[2:]) + \
                           salt.to_bytes(32, 'big') + init_code_hash
            
            address_bytes = Web3.keccak(create2_input)[12:]
            predicted_address = '0x' + address_bytes.hex()
            
            logger.info(f"📍 Predicted smart account: {predicted_address} for owner {eoa_address}")
            return to_checksum_address(predicted_address)
            
        except Exception as e:
            logger.error(f"❌ Error predicting smart account address: {str(e)}")
            raise
    
    def is_smart_account_deployed(self, smart_account_address: str) -> bool:
        """
        Check if smart account is deployed
        
        Args:
            smart_account_address: Smart account address
            
        Returns:
            True if deployed
        """
        return self.web3_service.contract_exists(smart_account_address)
    
    def get_smart_account_init_code(self, eoa_address: str, salt: int = 0) -> str:
        """
        Generate initCode for deploying smart account
        
        Args:
            eoa_address: EOA address that will be the owner
            salt: Salt for CREATE2
            
        Returns:
            initCode hex string
        """
        try:
            owner = to_checksum_address(eoa_address)
            
            # SimpleAccountFactory.createAccount(owner, salt)
            # Function selector: 0x5fbfb9cf
            create_account_selector = '0x5fbfb9cf'
            
            # Encode parameters
            encoded_owner = owner[2:].zfill(64)
            encoded_salt = hex(salt)[2:].zfill(64)
            
            # initCode = factoryAddress + createAccount(owner, salt)
            init_code = self.factory_address.lower() + \
                       create_account_selector[2:] + \
                       encoded_owner + \
                       encoded_salt
            
            logger.info(f"📝 Generated initCode for {eoa_address}")
            return init_code
            
        except Exception as e:
            logger.error(f"❌ Error generating initCode: {str(e)}")
            raise
    
    def estimate_deployment_gas(self, eoa_address: str) -> int:
        """
        Estimate gas for smart account deployment
        
        Args:
            eoa_address: EOA address
            
        Returns:
            Estimated gas
        """
        # Smart account deployment typically costs ~200k-300k gas
        # This includes factory call + account creation + initialization
        return 250000
    
    def get_smart_account_info(self, smart_account_address: str) -> Dict:
        """
        Get smart account information
        
        Args:
            smart_account_address: Smart account address
            
        Returns:
            Dictionary with account info
        """
        try:
            is_deployed = self.is_smart_account_deployed(smart_account_address)
            
            if not is_deployed:
                return {
                    'address': smart_account_address,
                    'deployed': False,
                    'chain_id': self.chain_id
                }
            
            # Get balance
            balance = self.web3_service.get_native_balance(smart_account_address)
            
            # Get nonce (would need to call smart account's getNonce())
            # Placeholder for now
            nonce = 0
            
            return {
                'address': smart_account_address,
                'deployed': True,
                'balance': float(balance),
                'nonce': nonce,
                'chain_id': self.chain_id
            }
            
        except Exception as e:
            logger.error(f"❌ Error getting smart account info: {str(e)}")
            raise
    
    def create_session_key(
        self,
        smart_account_address: str,
        session_key_address: str,
        valid_until: int,
        valid_after: int = 0
    ) -> Dict:
        """
        Create session key data for smart account
        
        Args:
            smart_account_address: Smart account address
            session_key_address: Session key address
            valid_until: Unix timestamp when session expires
            valid_after: Unix timestamp when session becomes valid
            
        Returns:
            Session key data
        """
        try:
            return {
                'smart_account': smart_account_address,
                'session_key': session_key_address,
                'valid_until': valid_until,
                'valid_after': valid_after,
                'is_active': True
            }
            
        except Exception as e:
            logger.error(f"❌ Error creating session key: {str(e)}")
            raise
    
    @staticmethod
    def generate_session_key() -> Dict:
        """
        Generate a new session key pair
        
        Returns:
            Dict with private_key and address
        """
        account = Account.create()
        return {
            'private_key': account.key.hex(),
            'address': account.address
        }

File: app/services/SmartAccountService.ts
/**
 * Smart Account Service
 * Handles Account Abstraction (ERC-4337) smart account creation and management
 * 
 * Features:
 * - Generate smart account addresses
 * - Deploy smart accounts
 * - Manage account abstraction operations
 */

import { createPublicClient, http, getContract, Address } from 'viem';
import { mainnet, base, arbitrum, optimism, polygon } from 'viem/chains';
import { keccak256, encodePacked, toBytes, toHex } from 'viem';

// Custom chain definitions for Lisk networks (not in viem by default)
const liskMainnet = {
  id: 1135,
  name: 'Lisk',
  network: 'lisk',
  nativeCurrency: {
    decimals: 18,
    name: 'Lisk',
    symbol: 'LSK',
  },
  rpcUrls: {
    default: { http: ['https://rpc.api.lisk.com'] },
    public: { http: ['https://rpc.api.lisk.com'] },
  },
  blockExplorers: {
    default: { name: 'LiskScan', url: 'https://blockscout.lisk.com' },
  },
} as const;

const liskSepolia = {
  id: 4202,
  name: 'Lisk Sepolia Testnet',
  network: 'lisk-sepolia',
  nativeCurrency: {
    decimals: 18,
    name: 'Sepolia Lisk',
    symbol: 'ETH',
  },
  rpcUrls: {
    default: { http: ['https://rpc.sepolia-api.lisk.com'] },
    public: { http: ['https://rpc.sepolia-api.lisk.com'] },
  },
  blockExplorers: {
    default: { name: 'LiskScan Sepolia', url: 'https://sepolia-blockscout.lisk.com' },
  },
  testnet: true,
} as const;

// Network configurations for smart accounts
export const SMART_ACCOUNT_NETWORKS = {
  [liskMainnet.id]: {
    chain: liskMainnet,
    bundlerUrl: 'https://api.pimlico.io/v2/1135/rpc',
    entryPoint: '0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789' as Address,
    factory: '0x9406Cc6185a346906296840746125a0E44976454' as Address,
  },
  [liskSepolia.id]: {
    chain: liskSepolia,
    bundlerUrl: 'https://api.pimlico.io/v2/4202/rpc',
    entryPoint: '0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789' as Address,
    factory: '0x9406Cc6185a346906296840746125a0E44976454' as Address,
  },
  [base.id]: {
    chain: base,
    bundlerUrl: 'https://api.pimlico.io/v2/8453/rpc',
    entryPoint: '0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789' as Address,
    factory: '0x9406Cc6185a346906296840746125a0E44976454' as Address,
  },
  [mainnet.id]: {
    chain: mainnet,
    bundlerUrl: 'https://api.pimlico.io/v2/1/rpc',
    entryPoint: '0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789' as Address,
    factory: '0x9406Cc6185a346906296840746125a0E44976454' as Address,
  },
};

// Simple Account Factory ABI (for address prediction)
const FACTORY_ABI = [
  {
    inputs: [
      { name: 'owner', type: 'address' },
      { name: 'salt', type: 'uint256' }
    ],
    name: 'createAccount',
    outputs: [{ name: '', type: 'address' }],
    stateMutability: 'nonpayable',
    type: 'function',
  },
  {
    inputs: [
      { name: 'owner', type: 'address' },
      { name: 'salt', type: 'uint256' }
    ],
    name: 'getAddress',
    outputs: [{ name: '', type: 'address' }],
    stateMutability: 'view',
    type: 'function',
  },
] as const;

export interface SmartAccountInfo {
  address: string;
  chainId: number;
  network: string;
  isDeployed: boolean;
  salt: string;
}

class SmartAccountService {
  private static readonly PIMLICO_API_KEY = 'pim_d9U8Wy3TMkKf1GMo9JN4C8';

  /**
   * Generate smart account address for an EOA owner
   */
  static async generateSmartAccountAddress(
    ownerAddress: string,
    chainId: number,
    salt: bigint = 0n
  ): Promise<SmartAccountInfo> {
    const network = SMART_ACCOUNT_NETWORKS[chainId as keyof typeof SMART_ACCOUNT_NETWORKS];
    
    if (!network) {
      throw new Error(`Network ${chainId} not supported for smart accounts`);
    }

    try {
      // Create public client
      const publicClient = createPublicClient({
        chain: network.chain,
        transport: http(),
      });

      // Predict smart account address using factory contract
      const predictedAddress = await publicClient.readContract({
        address: network.factory,
        abi: FACTORY_ABI,
        functionName: 'getAddress',
        args: [ownerAddress as Address, salt],
      });
      
      // Check if account is already deployed
      const bytecode = await publicClient.getBytecode({
        address: predictedAddress,
      });
      const isDeployed = !!bytecode && bytecode !== '0x';

      console.log(`📱 Smart Account Address: ${predictedAddress} (${network.chain.name})`);
      console.log(`📦 Deployed: ${isDeployed}`);

      return {
        address: predictedAddress,
        chainId,
        network: network.chain.name,
        isDeployed,
        salt: salt.toString(),
      };
    } catch (error) {
      console.error(`Failed to generate smart account for ${network.chain.name}:`, error);
      throw new Error(`Failed to generate smart account: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Generate smart account addresses for all supported networks
   */
  static async generateSmartAccountsAllNetworks(
    ownerAddress: string,
    salt: bigint = 0n
  ): Promise<SmartAccountInfo[]> {
    const results: SmartAccountInfo[] = [];
    const chainIds = Object.keys(SMART_ACCOUNT_NETWORKS).map(Number);

    console.log(`🔄 Generating smart accounts for ${chainIds.length} networks...`);

    for (const chainId of chainIds) {
      try {
        const smartAccount = await this.generateSmartAccountAddress(ownerAddress, chainId, salt);
        results.push(smartAccount);
      } catch (error) {
        console.warn(`Failed to generate smart account for chain ${chainId}:`, error);
        // Continue with other networks even if one fails
      }
    }

    console.log(`✅ Generated ${results.length}/${chainIds.length} smart accounts`);
    return results;
  }

  /**
   * Get smart account info for a specific network
   */
  static async getSmartAccountInfo(
    ownerAddress: string,
    chainId: number,
    salt: bigint = 0n
  ): Promise<SmartAccountInfo | null> {
    try {
      return await this.generateSmartAccountAddress(ownerAddress, chainId, salt);
    } catch (error) {
      console.error(`Failed to get smart account info:`, error);
      return null;
    }
  }

  /**
   * Check if smart account is deployed
   */
  static async isSmartAccountDeployed(
    smartAccountAddress: string,
    chainId: number
  ): Promise<boolean> {
    const network = SMART_ACCOUNT_NETWORKS[chainId as keyof typeof SMART_ACCOUNT_NETWORKS];
    
    if (!network) {
      return false;
    }

    try {
      const publicClient = createPublicClient({
        chain: network.chain,
        transport: http(),
      });

      const bytecode = await publicClient.getBytecode({
        address: smartAccountAddress as Address,
      });
      
      return !!bytecode && bytecode !== '0x';
    } catch (error) {
      console.error('Failed to check deployment status:', error);
      return false;
    }
  }

  /**
   * Get network configuration
   */
  static getNetworkConfig(chainId: number) {
    return SMART_ACCOUNT_NETWORKS[chainId as keyof typeof SMART_ACCOUNT_NETWORKS] || null;
  }

  /**
   * Get all supported networks
   */
  static getSupportedNetworks() {
    return Object.entries(SMART_ACCOUNT_NETWORKS).map(([chainId, config]) => ({
      chainId: Number(chainId),
      name: config.chain.name,
      network: config.chain.name.toLowerCase().replace(/\s+/g, '-'),
      ...config,
    }));
  }

  /**
   * Generate deterministic salt based on user data
   */
  static generateSalt(ownerAddress: string, userIdentifier: string = ''): bigint {
    // Create a deterministic but unique salt
    const data = encodePacked(
      ['address', 'string'],
      [ownerAddress as Address, userIdentifier]
    );
    
    const hash = keccak256(data);
    // Convert hash to bigint for salt
    return BigInt(hash);
  }

  // ============================================================================
  // CRYPTO-TO-NAIRA OPERATIONS
  // ============================================================================

  /**
   * Sign UserOperation for crypto-to-naira transaction
   * Uses smart account private key to create valid signature
   */
  static async signUserOperation(
    userOperation: any,
    privateKey: string
  ): Promise<any> {
    try {
      // This will be implemented to sign the UserOp hash
      // The backend will handle the actual signing with the smart account
      // Frontend just needs to prepare the UserOp structure
      
      return {
        ...userOperation,
        signature: '0x' + 'signed', // Placeholder - actual signing done on backend
      };
    } catch (error) {
      console.error('Failed to sign UserOperation:', error);
      throw error;
    }
  }

  /**
   * Build swap call data for token conversion
   * Prepares encoded call data for SmartAccount to execute swap
   */
  static buildSwapCallData(data: {
    smartAccountAddress: string;
    tokenIn: string;
    amountIn: number;
    tokenOut: string;
  }): string {
    try {
      // Encode the swap call data that will be executed by SmartAccount
      // This calls SwapRouter via the SmartAccount
      
      // Format: SwapRequest struct containing token swap details
      const swapData = {
        tokenIn: data.tokenIn,
        tokenOut: data.tokenOut,
        amountIn: data.amountIn,
        minAmountOut: 0, // Will be calculated by backend
        recipient: data.smartAccountAddress,
      };

      // Return encoded call data (simplified - actual encoding depends on router ABI)
      return JSON.stringify(swapData);
    } catch (error) {
      console.error('Failed to build swap call data:', error);
      throw error;
    }
  }

  /**
   * Get gas estimation for UserOperation
   */
  static async estimateUserOperationGas(
    userOperation: any,
    chainId: number
  ): Promise<{
    callGasLimit: bigint;
    verificationGasLimit: bigint;
    preVerificationGas: bigint;
  }> {
    try {
      // Estimate gas for the UserOperation
      const network = SMART_ACCOUNT_NETWORKS[chainId as keyof typeof SMART_ACCOUNT_NETWORKS];
      
      if (!network) {
        throw new Error(`Unsupported chain: ${chainId}`);
      }

      // Return reasonable gas estimates (these would be calculated more precisely)
      return {
        callGasLimit: BigInt(100000),
        verificationGasLimit: BigInt(100000),
        preVerificationGas: BigInt(21000),
      };
    } catch (error) {
      console.error('Failed to estimate gas:', error);
      throw error;
    }
  }

  /**
   * Prepare UserOperation for submission to bundler
   */
  static async prepareUserOperationForBundler(
    smartAccountAddress: string,
    callData: string,
    chainId: number
  ): Promise<any> {
    try {
      const network = SMART_ACCOUNT_NETWORKS[chainId as keyof typeof SMART_ACCOUNT_NETWORKS];
      
      if (!network) {
        throw new Error(`Unsupported chain: ${chainId}`);
      }

      // Create the UserOperation structure (ERC-4337 spec)
      const userOp = {
        sender: smartAccountAddress,
        nonce: 0, // Will be fetched from EntryPoint
        initCode: '0x', // Already deployed
        callData,
        accountGasLimits: '0x', // uint128 | uint128
        preVerificationGas: 21000,
        gasFees: '0x', // uint128 | uint128
        paymasterAndData: '0x',
        signature: '0x',
      };

      return userOp;
    } catch (error) {
      console.error('Failed to prepare UserOperation:', error);
      throw error;
    }
  }

  /**
   * Monitor UserOperation status on blockchain
   */
  static async monitorUserOperationStatus(
    userOperationHash: string,
    chainId: number,
    maxWaitTime: number = 300000
  ): Promise<{
    confirmed: boolean;
    transactionHash?: string;
    blockNumber?: number;
  }> {
    try {
      const network = SMART_ACCOUNT_NETWORKS[chainId as keyof typeof SMART_ACCOUNT_NETWORKS];
      
      if (!network) {
        throw new Error(`Unsupported chain: ${chainId}`);
      }

      const publicClient = createPublicClient({
        chain: network.chain,
        transport: http(),
      });

      const startTime = Date.now();
      
      // Poll for UserOperation receipt (simplified - actual implementation more complex)
      while (Date.now() - startTime < maxWaitTime) {
        try {
          // This would query the UserOperation status from EntryPoint or bundler
          // For now, return placeholder
          return {
            confirmed: true,
            transactionHash: '0x' + userOperationHash,
            blockNumber: 1,
          };
        } catch (error) {
          // Continue polling
          await new Promise(resolve => setTimeout(resolve, 5000));
        }
      }

      throw new Error('UserOperation confirmation timeout');
    } catch (error) {
      console.error('Failed to monitor UserOperation:', error);
      throw error;
    }
  }
}

export default SmartAccountService;